### ISSUE TITLE
twisted/trial/runner.py:243: exceptions.DeprecationWarning: log.startKeepingErrors is deprecated since Twisted 2.5

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2173]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2173|
|Type|defect|
|Created|2006-10-17 18:30:30Z|

Trial shouldn't cause this warning to be emitted.

<details><summary>Searchable metadata</summary>

```
trac-id__2173 2173
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1161109830000000 1161109830000000
changetime__1164103048000000 1164103048000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

Ready for review in startKeepingErrors-2173

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Generally looks good.  There are issues on Python 2.3, though.

```
exarkun@charm:/tmp$ PYTHONPATH=~/Projects/Twisted/trunk/ python2.3 -i ~/Projects/Twisted/trunk/bin/trial twisted
/home/exarkun/Projects/Twisted/trunk/twisted/conch/ssh/common.py:20: RuntimeWarning: PyCrypto not installed, but continuing anyways!
  RuntimeWarning)
/home/exarkun/Projects/Twisted/trunk/twisted/test/test_persisted.py:161: DeprecationWarning: twisted.persisted.marmalade is deprecated
  from twisted.persisted import marmalade
/home/exarkun/Projects/Twisted/trunk/twisted/words/im/__init__.py:8: UserWarning: twisted.im will be undergoing a rewrite at some point in the future.
  warnings.warn("twisted.im will be undergoing a rewrite at some point in the future.")
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/bin/trial", line 24, in ?
    run()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/scripts/trial.py", line 341, in run
    test_result = trialRunner.run(suite)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 734, in run
    suite = TrialSuite([test])
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 246, in __init__
    super(TrialSuite, self).__init__(suite)
  File "/usr/lib/python2.3/unittest.py", line 362, in __init__
    self.addTests(tests)
  File "/usr/lib/python2.3/unittest.py", line 379, in addTests
    for test in tests:
TypeError: iteration over non-sequence
>>> import pdb; pdb.pm()
> /usr/lib/python2.3/unittest.py(379)addTests()
-> for test in tests:
(Pdb) p type(tests)
<class 'twisted.trial.runner.LoggedSuite'>
(Pdb) 
```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to @exarkun|
|-|-|

Oops. Fixed. Running py2.3 build now. Submitting for re-review

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @jml|
|-|-|

Cool.  Merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set owner to @exarkun|
|-|-|

Oops. `TestSuite`s aren't acceptable arguments for `addError`.  I did the right thing here, so it'll probably need a re-review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jerub|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_8"></a>|@jerub set owner to @jml|
|-|-|

When viewing this diff, I found this command useful:
```
$ svn diff --diff-cmd "diff" --extensions "-ubB" | less
```

These pyflakes warnings should be cleaned up. You're importing TrialRunner but using runner.TrialRunner
```
twisted/trial/test/test_runner.py:92: 'TrialRunner' imported but unused
twisted/trial/test/test_runner.py:289: 'TrialSuite' imported but unused
```

I'm not sure what to think about the lack of docstrings on setUp and tearDown methods. Nor on some of the more trivial methods, such as 'def id(self): return self.description'.

Reading this code further, I've discovered that ITestCase, twisted.trial.unittest.TestCase both do not implement or document the 'id()' method, but it's called in t.t.ut.TestCase.run(). Where does it come from? Why is it not documented?

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_9"></a>|@jerub commented|
|-|-|

We discussed this on irc, the pyflakes warnings were cleaned up and docstrings were promised. on TestErrorHolder.setUp and LoggedSuite.run. Once those are added, this can be merged.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml set status to `closed`|
|-|-|

(In [18764]) Use `LoggedSuite`  instead of `startKeepingErrors` to trap logged errors
 * Author: jml
 * Reviewers: Jerub, exarkun
 * Fixes [#2173](https://github.com/twisted/twisted/issues/2173)

`startKeepingErrors` has been deprecated for a while, but Trial was still
using it to catch errors that were logged during a test run, but outside 
of any particular test case. 

This branch creates a new kind of `TestSuite` called `LoggedSuite`, which
monitors the log and adds any logged errors to the test result object.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

(In [18767]) Use `LoggedSuite`  instead of `startKeepingErrors` to trap logged errors
 * Author: jml
 * Reviewers: Jerub, exarkun
 * Fixes [#2173](https://github.com/twisted/twisted/issues/2173)
 * Remerged with fix for test regression caused by r18764

`startKeepingErrors` has been deprecated for a while, but Trial was still
using it to catch errors that were logged during a test run, but outside
of any particular test case.

This branch creates a new kind of `TestSuite` called `LoggedSuite`, which
monitors the log and adds any logged errors to the test result object.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
